### DISCUSSION TITLE
How are the various starter config files related with one another in terms of the starters?

### DISCUSSION THREAD

**Discussion by simkimsia** [kedro-org/kedro#1186] [Idea]
I'm reading this https://kedro.readthedocs.io/en/stable/07_extend_kedro/05_create_kedro_starters.html#configuration-variables

and i like to ask how are

1. config.yml
2. prompts.yml
3. and the proto-typical cookiecutter.json related with one another?

And what are the key functions that are associated with manipulating input from user via CLI or from cookiecutter.json/config.yml and then pass to cookiecutter?

**Comment by antonymilne** [comment]
Hi @simkimsia, great question. Let me explain them in the opposite order to how you listed them since I think it's easiest to understand that way.

[cookiecutter.json](https://github.com/kedro-org/kedro/blob/main/kedro/templates/project/cookiecutter.json) is something from the original cookiecutter package. It provides a basic mechanism to render prompts and receive user input on the CLI, passing the received input into the project template and replacing placeholders like `{{ cookiecutter.project_name }}`. cookiecutter.json gives you a way to specify default values for variables but not much else.

[prompts.yml](https://github.com/kedro-org/kedro/blob/main/kedro/templates/project/prompts.yml) is kedro-specific. It adds extra functionality to the prompts, e.g. regex validation and explanatory text. You can see it like an extension of cookiecutter.json that adds functionality not there in cookiecutter itself.

config.yml is a _user-specified_ input file which can be used as an alternative to the interactive prompt. This is useful if you wish to create a kedro project programatically and don't want to go through the process of manually entering values. e.g. the following file
```
# config.yml
project_name: Something
repo_name: something
python_package: something
```

can be used with `kedro new --config=config.yml`. The user won't be prompted to enter any values for the cookiecutter variables, since the values have been specified already in the config file. Note that calling the file config.yml is just a convention here; you could call it anything you like, so long as it's a valid yaml file.

If you want to see the code that does this, it's all in [starters.py](https://github.com/kedro-org/kedro/blob/main/kedro/framework/cli/starters.py). I rewrote this file fairly recently so hopefully it's reasonably easy to follow now.

**Reply by datajoely** [reply]
Whilst not related to the cookiecutter content @AntonyMilneQB explains - for anyone else arriving here looking for wider explanation on Kedro configuration beyond [what we have in the docs](https://kedro.readthedocs.io/en/latest/04_kedro_project_setup/02_configuration.html) - this image is our current best description of the 'levels' and responsibilities in respect to a Kedro run:

![image](https://user-images.githubusercontent.com/35801847/151771433-3654bfc9-ea8c-4654-90e4-a4dd14db3365.png)

**Reply by simkimsia** [reply]
@AntonyMilneQB oh thank you. great answer.

By the way, i recently learned from kedro docs that a Kedro starter is essentially a cookiecutter template.

My followup questions are:

1. when a user `kedro new` which starter in kedro-starters does kedro use? i have been trying to read the code, but i haven't gotten round to that `def new` yet. i just figured out `def info` the other day. i saw kedro-starters as a github repo.
2. when a starter vendor (the party responsible for creating a kedro starter) creates a starter to be used, do they need to do BOTH the prompts.yml and the config.yml? or just the prompts.yml? Where do they place their default values? inside cookiecutter.json or they supply config.yml expecting it to be overridden by a starter user (the party that's using the starter to generate a new project)?

Thank you 🙏

**Reply by simkimsia** [reply]
@datajoely sorry but what's a Kedro Run? Is it when somebody uses kedro to create a new project? or when they "run" a project that was generated from Kedro CLI? 

I like the image. i wish there's a similar one that explains what @AntonyMilneQB did regarding the configs. so that newbies can quickly learn the different configs and how they interact with one another

**Reply by datajoely** [reply]
Hi @simkimsia 

- The default Kedro template can be reviewed [here](https://github.com/kedro-org/kedro/tree/main/kedro/templates/project). 
- We have our tutorial on how to make your own starter [here](https://kedro.readthedocs.io/en/stable/07_extend_kedro/05_create_kedro_starters.html). 
- Quickly looking through `prompts.yml` is definitely required, as is `cookiecutter.json` - but I'm not sure you need anything called `config.yml`. Kedro itself needs some stuff in it's `conf` folder, but that's not related to cookiecutter.

**Reply by datajoely** [reply]
And a `kedro run` is when you execute the pipelines not the creation of a starter - I posted that because I suspect people arriving at this discussion may not realise this is specifically to do with starters

**Reply by simkimsia** [reply]
Got it. So default values are in cookiecutter.json okay I get it now.

**Reply by simkimsia** [reply]
@datajoely to allay your concerns abt the possible confusion abt config files for running pipelines I have altered the title. Hope that helps if not let me know and I will change again
